test/e2e/subscription_e2e_test.go

```diff
@@ -1265,6 +1265,145 @@ var _ = Describe("Subscription", func() {
+		csvA := newCSV("nginx-a", testNamespace, "", semver.MustParse("0.1.0"), nil, []apiextensions.CustomResourceDefinition{crd, crd2}, namedStrategy)
+		csvAB := newCSV("nginx-a-b", testNamespace, "", semver.MustParse("0.1.0"), []apiextensions.CustomResourceDefinition{crd}, nil, namedStrategy)
+		csvB := newCSV("nginx-b-dep", testNamespace, "", semver.MustParse("0.1.0"), []apiextensions.CustomResourceDefinition{crd}, nil, depNamedStrategy)
+		csvC := newCSV("nginx-c-dep", testNamespace, "", semver.MustParse("0.1.0"), []apiextensions.CustomResourceDefinition{crd2}, nil, depNamedStrategy2)
+
+		// Create PackageManifests
+		manifests := []registry.PackageManifest{
+			{
+				PackageName: packageName1,
+				Channels: []registry.PackageChannel{
+					{Name: stableChannel, CurrentCSVName: csvA.GetName()},
+					{Name: alphaChannel, CurrentCSVName: csvAB.GetName()},
+				},
+				DefaultChannelName: stableChannel,
+			},
+			{
+				PackageName: packageName2,
+				Channels: []registry.PackageChannel{
+					{Name: stableChannel, CurrentCSVName: csvB.GetName()},
+				},
+				DefaultChannelName: stableChannel,
+			},
+		}
+
+		manifests2 := []registry.PackageManifest{
+			{
+				PackageName: packageName2,
+				Channels: []registry.PackageChannel{
+					{Name: stableChannel, CurrentCSVName: csvC.GetName()},
+				},
+				DefaultChannelName: stableChannel,
+			},
+		}
+
+		catalogSourceName := genName("catsrc")
+		catsrc, cleanup := createInternalCatalogSource(GinkgoT(), kubeClient, crClient, catalogSourceName, testNamespace, manifests, []apiextensions.CustomResourceDefinition{crd, crd2}, []v1alpha1.ClusterServiceVersion{csvA, csvAB, csvB})
+		require.NotNil(GinkgoT(), catsrc)
+		require.NotNil(GinkgoT(), cleanup)
+		defer cleanup()
+
+		// Ensure that the catalog source is resolved before we create a subscription.
+		_, err := fetchCatalogSource(GinkgoT(), crClient, catsrc.GetName(), testNamespace, catalogSourceRegistryPodSynced)
+		require.NoError(GinkgoT(), err)
+
+		subscriptionSpec := &v1alpha1.SubscriptionSpec{
+			CatalogSource:          catsrc.GetName(),
+			CatalogSourceNamespace: catsrc.GetNamespace(),
+			Package:                packageName1,
+			Channel:                stableChannel,
+			StartingCSV:            csvA.GetName(),
+			InstallPlanApproval:    v1alpha1.ApprovalAutomatic,
+		}
+
+		catalogSourceName2 := genName("catsrc")
+		catsrc2, cleanup2 := createInternalCatalogSource(GinkgoT(), kubeClient, crClient, catalogSourceName2, testNamespace, manifests2, []apiextensions.CustomResourceDefinition{crd2}, []v1alpha1.ClusterServiceVersion{csvC})
+		require.NotNil(GinkgoT(), catsrc)
+		require.NotNil(GinkgoT(), cleanup)
+		defer cleanup2()
+
+		// Ensure that the catalog source is resolved before we create a subscription.
+		_, err = fetchCatalogSource(GinkgoT(), crClient, catsrc2.GetName(), testNamespace, catalogSourceRegistryPodSynced)
+		require.NoError(GinkgoT(), err)
+
+		// Create a subscription that has a dependency
```

Comment:
Hey Harish. Thank you for the feedbacks on the e2e test case. This PR is designed to be able to backport all the way down to 4.2. As a result, I'm sticking to "vanilla" version of e2e test structure so that it can be backported easier. There will be a follow-up PR that will clean all of these up in master later. In fact, there are a big chuck of code here which are only here for backport reason and will be removed after the backport process done along side with the cleanup PR.